### PR TITLE
8285792: Posix signal handler modification checking issues.

### PR DESCRIPTION
**Pull request by hseigel**
Please review this fix for JDK-8285792.  The fix removes print statements from check_signal_handler() so that it doesn't print all the handlers every time it finds one that is modified.  Instead, it returns true if the handler is modified, false otherwise.  Its caller, user_handler(), then prints all the handlers just once even if multiple signal handlers were modified.

The fix also adds a check for VMError::crash_handler_adress() to check_signal_handler() to prevent it from being treated as a signal handler modification.

The fix was tested with Mach5 tiers 1-2 on Linux and Mac OS and Mach 5 tiers 3-5 on Linux x64.  The regression test is excluded on Windows.

Thanks, Harold

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8285792](https://bugs.openjdk.org/browse/JDK-8285792): Posix signal handler modification checking issues.

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9631/head:pull/9631` \
`$ git checkout pull/9631`

Update a local copy of the PR: \
`$ git checkout pull/9631` \
`$ git pull https://git.openjdk.org/jdk pull/9631/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9631`

View PR using the GUI difftool: \
`$ git pr show -t 9631`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9631.diff">https://git.openjdk.org/jdk/pull/9631.diff</a>

</details>

### CODE REVIEW

**Comment by hseigel** [comment]
I ran the test for JDK-8279878 after removing both this fix and the fix for JDK-8279878 and the test still passed.  So I don't think its failure was related to this issue.  This fix prevents the output from occurring more than once.  It does not prevent it from occurring at all.

The JVM was reporting that it found an unexpected handler for SIG_IGN.  Is it possible that such a handler is no longer being defined?

Note that the JVM only checks for unexpected signal handlers if -Xcheck:jni or -XX:+CheckJNICall is specified.

**Comment by prrace** [comment]
> I ran the test for JDK-8279878 after removing both this fix and the fix for JDK-8279878 and the test still passed. So I don't think its failure was related to this issue. This fix prevents the output from occurring more than once. It does not prevent it from occurring at all.
> 
> The JVM was reporting that it found an unexpected handler for SIG_IGN. Is it possible that such a handler is no longer being defined?
> 
> Note that the JVM only checks for unexpected signal handlers if -Xcheck:jni or -XX:+CheckJNICall is specified.

Sure .. we updated the test to ignore these warnings ... but I was hoping that maybe your changes would obsolete the need to ignore them. However I actually suspect it won't help if it just minimises them rather than eliminates them.

It stems from the GTK library (the standard library for Gnome desktop apps on Linux) calling 
(SIGPIPE, SIG_IGN) during initialization

Their reasons are here : https://docs.gtk.org/gtk3/func.init.html

So the VM warnings are something of a mild annoyance that we've worked around in this test but it never made sense to me what this has to do with JNI ?

File: src/hotspot/os/posix/signals_posix.cpp [hunk 1]
```diff
@@ -832,28 +833,28 @@ static void check_signal_handler(int sig) {
...
-    if (os_sigaction == NULL) return;
+    if (os_sigaction == NULL) return false;
   }
 
   os_sigaction(sig, (struct sigaction*)NULL, &act);
 
   // Compare both sigaction structures (intelligently; only the members we care about).
-  if (!are_handlers_equal(&act, expected_act)) {
+  // Ignore if the handler is our own crash handler.
+  if (!are_handlers_equal(&act, expected_act) &&
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/signals_posix.cpp:843]: Pre-existing nit: really this should be called `are_actions_equal`. The poor naming is clearer when you now follow this with  `HANDLER_IS`.

**Comment by dholmes-ora** [comment]
Looking at [JDK-4229104](https://bugs.openjdk.org/browse/JDK-4229104) I have to wonder whether our handling of `SIGPIPE` is still necessary? Even if so, the fact we always ignore it if we get it suggests we really don't care if someone else has installed a (temporary) handler for it, so perhaps we should not include it (or `SIGXFSZ`) in these checks?

File: test/hotspot/jtreg/runtime/posixSig/TestPsig.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+public class TestPsig {
+    static {
+        System.loadLibrary("TestPsig");
+    }
+    public static native void doSomething(int val);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/posixSig/TestPsig.java:28]: Suggestion:

`private static native void changeSigActionFor(int sig);`

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Harold,

Overall seems fine. A couple of nits.

We could simplify the test by just having (as is common) on Java file which launches itself in a new VM when invoked with no-args, and does the real test when passed a (dummy) arg. Then we will have one less file and clearer names.

Thanks.

**Comment by hseigel** [comment]
>> So the VM warnings are something of a mild annoyance that we've worked around in this test but it never made sense to me what this has to do with JNI ?

At some point it was decided that checkJNI should check for certain user defined signal handlers.  Changing that would require a CSR and is outside the scope of this pull request.

File: test/hotspot/jtreg/runtime/posixSig/TestPsig.java (hunk 2 above)
Comment (hseigel) [reply test/hotspot/jtreg/runtime/posixSig/TestPsig.java]: Fixed in latest commit.

File: src/hotspot/os/posix/signals_posix.cpp (hunk 1 above)
Comment (hseigel) [reply src/hotspot/os/posix/signals_posix.cpp]: Thanks for the suggestion.  Function are_handlers_equal() is renamed to are_actions_equal() in the latest commit.

**Comment by hseigel** [comment]
@dholmes-ora Thanks for looking at this.  Please review the latest changes.  Should I enter a new rfe to look at JVM handling of SIGPIPE and SIGXFSZ in these checks?

**Review by dholmes-ora — APPROVED** [review]
Hi Harold,

Updates look good - thanks.

Please file a RFE to look at omitting SIGPIPE and perhaps SIGXFSZ from the check.

**Review by erikj79 — APPROVED** [review]
Build change looks good.

**Comment by hseigel** [comment]
Thanks David and Erik for the reviews!

/integrate

**Pull request closed without merging**